crates/pallet-domains/src/staking.rs

```diff
@@ -1456,6 +1480,113 @@ pub(crate) fn do_mark_operators_as_slashed<T: Config>(
     Ok(())
 }
 
+pub(crate) fn do_mark_invalid_bundle_authors<T: Config>(
+    domain_id: DomainId,
+    er: &ExecutionReceiptOf<T>,
+) -> Result<(), Error> {
+    let invalid_bundle_authors = invalid_bundle_authors_for_receipt::<T>(domain_id, er);
+    let er_hash = er.hash::<DomainHashingFor<T>>();
+    let pending_slashes = PendingSlashes::<T>::get(domain_id).unwrap_or_default();
+    let mut stake_summary =
+        DomainStakingSummary::<T>::get(domain_id).ok_or(Error::DomainNotInitialized)?;
+
+    for operator_id in invalid_bundle_authors {
+        if pending_slashes.contains(&operator_id) {
+            continue;
+        }
+
+        mark_invalid_bundle_author::<T>(operator_id, er_hash, &mut stake_summary)?;
+    }
+
+    DomainStakingSummary::<T>::insert(domain_id, stake_summary);
+    Ok(())
+}
+
+fn mark_invalid_bundle_author<T: Config>(
+    operator_id: OperatorId,
+    er_hash: ReceiptHashFor<T>,
+    stake_summary: &mut StakingSummary<OperatorId, BalanceOf<T>>,
+) -> Result<(), Error> {
+    Operators::<T>::try_mutate(operator_id, |maybe_operator| {
+        let operator = match maybe_operator.as_mut() {
+            // If the operator is already slashed and removed due to fraud proof, when the operator
+            // is slash again due to invalid bundle, which happen after the ER is confirmed, we can
+            // not find the operator here thus just return.
+            None => return Ok(()),
+            Some(operator) => operator,
+        };
+
+        // operator must be in registered status.
+        // for other states, we anyway do not allow bundle submission.
+        if operator.status::<T>(operator_id) != &OperatorStatus::Registered {
+            return Ok(());
+        }
+
+        // slash and remove operator from next and current epoch set
+        operator.update_status(OperatorStatus::InvalidBundle(er_hash));
+        if stake_summary
+            .current_operators
+            .remove(&operator_id)
+            .is_some()
+        {
+            stake_summary.current_total_stake = stake_summary
+                .current_total_stake
+                .checked_sub(&operator.current_total_stake)
+                .ok_or(Error::BalanceUnderflow)?;
+        }
+        stake_summary.next_operators.remove(&operator_id);
+        Ok(())
+    })
+}
+
+pub(crate) fn do_unmark_invalid_bundle_authors<T: Config>(
+    domain_id: DomainId,
+    er: &ExecutionReceiptOf<T>,
+) -> Result<(), Error> {
+    let invalid_bundle_authors = invalid_bundle_authors_for_receipt::<T>(domain_id, er);
+    let er_hash = er.hash::<DomainHashingFor<T>>();
+    let pending_slashes = PendingSlashes::<T>::get(domain_id).unwrap_or_default();
+    let mut stake_summary =
+        DomainStakingSummary::<T>::get(domain_id).ok_or(Error::DomainNotInitialized)?;
+
+    for operator_id in invalid_bundle_authors {
+        if pending_slashes.contains(&operator_id) {
```

Comment:
Should also check `is_operator_pending_to_slash` to see if an operator is pending slash.